### DISCUSSION TITLE
Conditionally initializing an array of nullable reference types

### DISCUSSION THREAD

**Discussion by foxesknow** [dotnet/csharplang#3406] [Q&A]
I'm not 100% sure if this is an issue, but thought I'd post it anyway just to be sure...

If you've got an array of nullable reference types that you conditionally initialize via an if statement then the compiler doesn't do the flow analysis to detect that the item is no longer null. For example:

```

private static List<string>?[] m_Buckets = new List<string>[10];

public static void Update(int index, string value)
{
    if(m_Buckets[index] is null)
    {
        m_Buckets[index] = new List<string>();
    }

    // bucket is List<string>?
    var bucket = m_Buckets[index];
}
```

If this an issue with the flow analysis in the compiler, or is it a conscious decision to not analyze this sort of flow?

**Comment by ufcpp** [comment]
how about

```cs
var bucket = m_Buckets[index] ??= new List<string>();
```

**Comment by foxesknow** [comment]
> 
> 
> how about
> 
> ```cs
> var bucket = m_Buckets[index] ??= new List<string>();
> ```

Yep, that works!

However, I'm curious if my example should work in the first place?

**Comment by YairHalberstadt** [comment]
This scenario is considered too complex for the nullable flow state analyzer to handle at the moment. However it's possible that may change in the future.
